### PR TITLE
[FLINK-11486][tests] Port RecoveryITCase to new code base

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

Port `RecoveryITCase` to new code base.

This PR is based on #7676.

## Brief change log

- "recover a task manager failure" --> TaskExecutorITCase#testJobRecoveryWithFailingTaskExecutor
- "recover once failing forward job" --> JobRecoveryITCase#testTaskFailureRecovery
- "recover once failing forward job with slot sharing" --> JobRecoveryITCase#testTaskFailureWithSlotSharingRecovery

## Verifying this change

- Executed ported tests repeatedly

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java [hunk 1]
```diff
@@ -78,22 +81,14 @@ public void teardown() throws Exception {
 	}
 
 	/**
-	 * Tests that a job will be re-executed if a new TaskExecutor joins the cluster.
+	 * Tests that a job can be re-executed after the job has failed due
+	 * to a TaskExecutor termination.
 	 */
 	@Test
-	public void testNewTaskExecutorJoinsCluster() throws Exception {
+	public void testCleanClusterStateAfterTaskExecutorTermination() throws Exception {
```
Comment (Clarkkkkk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java]: The name is not very intuitive, I think testJobReExecutionAfterTaskExecutorTermination would be better.

Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java]: Good idea. I will apply the changes.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java [hunk 2]
```diff
@@ -111,6 +106,40 @@ public void testNewTaskExecutorJoinsCluster() throws Exception {
...
+		miniCluster.submitJob(jobGraph).get();
+
+		final CompletableFuture<JobResult> jobResultFuture = miniCluster.requestJobResult(jobGraph.getJobID());
+
+		assertThat(jobResultFuture.isDone(), is(false));
+
+		CommonTestUtils.waitUntilCondition(
+			jobIsRunning(() -> miniCluster.getExecutionGraph(jobGraph.getJobID())),
+			Deadline.fromNow(TESTING_TIMEOUT),
+			20L);
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java]: There is a bug in `waitUntilCondition` which leads to `20` being ignored.

Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java]: Good catch. Will fix it while merging this PR.

**Comment by tillrohrmann** [comment]
Thanks for the review @GJL. I'll address your comment while merging this PR.

**Pull request closed without merging**